### DISCUSSION TITLE
Get weights and bias from the saved model

### DISCUSSION THREAD

**Discussion by wangjiawen2013** [tracel-ai/burn#2402] [Q&A] [answered]
Hi,
where we can find the weights and bias from the saved burn model ?

**Comment by laggui** [comment, marked as answer]
The parameters for each module are accessible via their respective field. For example, you can get the `Linear` weight and bias tensor values with:

```rust
let weight = linear.weight.val();
let bias = linear.bias.unwrap().val();
```

**Comment by wangjiawen2013** [comment]
@laggui 
Thanks, and this is a simple case. However, things become complicated in my case.
I construct a model consisting of many blocks, and there are many various layers in each block, such as linear, conv2d and leakyrelu and so. And I want to clip the parameters. 

In summary, in a complicated nested model, how to get and modify the parameters in a convenient way ?

With pytorch, this can be achieved by the following code:
```python
 for p in discriminator.parameters():
    print(p.data.type)
    p.data.clamp(-clip_value, clip_value)
```

**Reply by laggui** [reply]
Ahhh ok in this case you probably want to use a [`ModuleMapper`](https://burn.dev/burn-book/building-blocks/module.html#visitor--mapper) similar to what is implemented for the [optimizer adapter](https://github.com/tracel-ai/burn/blob/main/crates/burn-core/src/optim/simple/adaptor.rs#L113) which includes gradient clipping (though you're not dealing with gradients, just the values themselves in your case).

Each float parameter will go through this mapping so you can simply apply your clipping there.

**Comment by wangjiawen2013** [comment]
@laggui 
Hi,
I am troubling you again!
I am trying to hard to under the working mechanism of `ModuleMapper` but failed to understand what it is. This is the definition of some methods:
```rust
/// Module visitor trait.
pub trait ModuleVisitor<B: Backend> {
    /// Visit a tensor in the module.
    fn visit<const D: usize>(&mut self, id: ParamId, tensor: &Tensor<B, D>);
}

/// Module mapper trait.
pub trait ModuleMapper<B: Backend> {
    /// Map a tensor in the module.
    fn map<const D: usize>(&mut self, id: ParamId, tensor: Tensor<B, D>) ->
      Tensor<B, D>;
}
```
1) Whait's the parameter tensor in in `visit` and `map` does ? And what's the tensor that `map` returns do ? is it different from the tensor in the input parameter ?  I think we just want to visit a paramid, so why a tensor is needed in the `visit` parameter ?

2) How to get all the `ParamId` in a layerblock or model ?

3) Could you show a simple example on how to modify the parameters of a model ?

**Reply by laggui** [reply]
1. The `tensor` contains the floating-point tensor parameter, so if you want to modify its value (e.g., clip between a range) this is the tensor that you want to operate on. Since you want to modify its value, you should use the `ModuleMapper` as it allows you to take the original parameter tensor as input and return its modified value.
2. In order to get all the `ParamId` in a module you need to iterate over all the parameters to capture a tensor param's `ParamId`. If you don't care about mapping a specific parameter id to its module, then you could possibly implement the `ModuleVisitor` to fill a container with all the parameter ids of your module.
3. Here is a quick example which very conveniently clips all module parameters between a range of values.
```rust
use burn::{
    backend::NdArray,
    module::{Module, ModuleMapper},
    nn::{Linear, LinearConfig},
    prelude::Backend,
};

#[derive(Debug, Module)]
pub struct MyModel<B: Backend> {
    pub linear: Linear<B>,
}

pub struct Clip {
    pub min: f32,
    pub max: f32,
}

impl<B: Backend> ModuleMapper<B> for Clip {
    fn map_float<const D: usize>(
        &mut self,
        _id: burn::module::ParamId,
        tensor: burn::prelude::Tensor<B, D>,
    ) -> burn::prelude::Tensor<B, D> {
        tensor.clamp(self.min, self.max)
    }
}

fn main() {
    type B = NdArray;
    let device = Default::default();

    // Initial weights
    let model = MyModel {
        linear: LinearConfig::new(10, 2).init::<B>(&device),
    };
    println!("Before:\n{}\n", model.linear.weight.val());

    // Clip module mapper
    let mut clip = Clip {
        min: -0.01,
        max: 0.01,
    };
    let model = model.map(&mut clip);

    println!("After:\n{}\n", model.linear.weight.val());
}
```

**Reply by wangjiawen2013** [reply]
I reproduced this example successfully, however, when I use "clip" in my model, things became weired !  The parameters updated as expected when I didn't clip the parameters, but when I clip them, the parameters were not updated at all during the training !

This is the clip I used:
```rust
/// Clip module mapper to clip all module parameters between a range of values
#[derive(Module, Clone, Debug)]
pub struct Clip {
    pub min: f32,
    pub max: f32,
}

impl<B: Backend> ModuleMapper<B> for Clip {
    /// Map an float tensor in the module. Each ParamId and it's tensor value will be transfered
    /// to map_float and the modified value will be retured.
    fn map_float<const D: usize> (&mut self, _id: ParamId, tensor: Tensor<B, D>) -> Tensor <B, D> {
        tensor.clamp(self.min, self.max)
    }
}
```
And this is a chunk of code from the training.rs, the parameters were not updated at all !
```rust
let mut clip = Clip {min: -0.01, max: 0.01};

let loss_d = -discriminator.forward(batch.images).mean() + discriminator.forward(fake_images).mean();

// Gradients for the current backward pass
let grads = loss_d.backward();
// Gradients linked to each parameter of the discriminator
let grads = GradientsParams::from_grads(grads, &discriminator);
// Update the discriminator using the optimizer
discriminator = optimizer_d.step(config.lr, discriminator, grads);
// Clip parameters (weights) of discriminator
discriminator = discriminator.map(&mut clip);
 ```
When I commented out the last line, the parameters can be updated normally during the training iteration. I thinks this is perhaps the reason which caused the failure of my model (see this disscussion https://github.com/tracel-ai/burn/discussions/2593)

**Reply by laggui** [reply]
Ahhh ok it was not clear what the weight clipping was supposed to be for, but in this case you need to make sure that the `require_grad` attribute is handled correctly. Otherwise the resulting parameter might not be tracked.

This is similar to the [optimizer adapter](https://github.com/tracel-ai/burn/blob/main/crates/burn-core/src/optim/simple/adaptor.rs#L113) that I linked earlier in the discussion:

```rust
impl<B: AutodiffBackend> ModuleMapper<B> for Clip {
    fn map_float<const D: usize>(&mut self, _id: ParamId, tensor: Tensor<B, D>) -> Tensor<B, D> {
        let is_require_grad = tensor.is_require_grad();

        let mut tensor = Tensor::from_inner(tensor.inner().clamp(self.min, self.max));

        if is_require_grad {
            tensor = tensor.require_grad();
        }
        tensor
    }
}
```

It should fix the issue you're currently observing with wgan where the loss does not seem to really fluctuate (probably because parameters were never updated after the first time they were clipped).

**Reply by wangjiawen2013** [reply]
I have re-implemented wgan in burn, the performance is comparable with that of pytorch. I'll pull a request later.

**Comment by wangjiawen2013** [comment]
@laggui 
Thanks for your help ! I can clip parameters now. 
You said "In order to get all the ParamId in a module you need to iterate over all the parameters to capture a tensor param's ParamId.".  l read carefully on burn's docs and source code but still cannot find how to do it. So my question is:
1. How get all the paramid in a module.
2. How to modify a specific parameters (I don't need this but it may help later.).
A chunk of exemple code is prefered !

**Reply by laggui** [reply]
1. Here's an example to illustrate how to use the visitor to retrieve the parameter ids for your model.
```rust
use burn::{
    backend::NdArray,
    module::{Module, ModuleVisitor, ParamId},
    nn::{Linear, LinearConfig},
    prelude::Backend,
    tensor::Tensor,
};

#[derive(Debug, Module)]
pub struct MyModel<B: Backend> {
    pub linear: Linear<B>,
    pub layer: MyLayer<B>,
}

#[derive(Debug, Module)]
pub struct MyLayer<B: Backend> {
    pub l1: Linear<B>,
    pub l2: Linear<B>,
}

pub struct ParamContainer {
    pub ids: Vec<ParamId>,
}

impl ParamContainer {
    pub fn new() -> Self {
        Self { ids: Vec::new() }
    }
}

impl<B: Backend> ModuleVisitor<B> for ParamContainer {
    fn visit_float<const D: usize>(&mut self, id: ParamId, _tensor: &Tensor<B, D>) {
        self.ids.push(id);
    }
}

fn main() {
    type B = NdArray;
    let device = Default::default();

    let model: MyModel<B> = MyModel {
        linear: LinearConfig::new(10, 16).init(&device),
        layer: MyLayer {
            l1: LinearConfig::new(16, 8).init(&device),
            l2: LinearConfig::new(8, 2).init(&device),
        },
    };

    // Print the parameter IDs
    let ids = vec![
        model.linear.weight.id,
        model.linear.bias.clone().unwrap().id,
        model.layer.l1.weight.id,
        model.layer.l1.bias.clone().unwrap().id,
        model.layer.l2.weight.id,
        model.layer.l2.bias.clone().unwrap().id,
    ];
    println!("Module:\n{ids:?}");

    let mut container = ParamContainer::new();

    model.visit(&mut container);

    // Now we can see which parameter IDs are in the module
    println!("Parameters:\n{:?}", container.ids);

    // For sanity
    assert_eq!(ids, container.ids);
}
```

2. You can apply a mapping to a specific parameter, for example:
```rust
let model: MyModel<B> = MyModel {
    linear: LinearConfig::new(10, 16).init(&device),
    layer: MyLayer {
        l1: LinearConfig::new(16, 8).init(&device),
        l2: LinearConfig::new(8, 2).init(&device),
    },
};

// Apply a function to the weights (here, clamp between -0.5 and 0.5)
model.linear.weight.map(|x| x.clamp(-0.5, 0.5))
```

**Reply by wangjiawen2013** [reply]
In this example, all the fields in the structs are `pub`. This is  necessary because we must access these fields to get weights and bias. However, in burn's github examples, the fields in most of the structs are private by default. Should we use "public" or "private" fields when we use burn ?

Besides, `setters` and `getters` are used widely in some object-oriented languages, such as java and pytorch, but they are never used in burn's github examples. This makes some users confused because we don't know whether to use `setters` and `getters` in burn/rust. Do you have any suggestions or best practices ?

The third question is about the model definition. In burn's github examples, all the model structs are made of a sequence of modules, for example:
```rust
#[derive(Module, Debug)]
pub struct Model<B: Backend> {
    conv1: ConvBlock<B>,
    conv2: ConvBlock<B>,
    conv3: ConvBlock<B>,
    dropout: nn::Dropout,
    fc1: nn::Linear<B>,
    fc2: nn::Linear<B>,
    activation: nn::Gelu,
}
```
Can we define some non-module fields in the model ? for example if we define `dimension1`, `dimension2` and `normalize` in the model, then we can use these field(`self.dimension, self.dimension2, self.normalize`) in the model's `forward` method:
```rust
pub struct Model<B: Backend> {
    dimension1: usize,
    dimension2: usize,
    normalize: bool,
    conv1: ConvBlock<B>,
    conv2: ConvBlock<B>,
    conv3: ConvBlock<B>,
    dropout: nn::Dropout,
    fc1: nn::Linear<B>,
    fc2: nn::Linear<B>,
    activation: nn::Gelu,
}
```

**Reply by laggui** [reply]
1. Fields can be either private or public, it really depends on the implementation and usage. The visibility of a field is up to what you expect users to do with it.
2. Rust is not an object-oriented language, so getters and setters are not as prominent. If you want to modify or retrieve the value of a private field, you could use something like a getter or setter. Otherwise, the visibility of the field means that you can access and/or modify it. It can either be public, public for the current crate only, or private.
3. Yep, there should be no issue with that 🙂 though these are treated as constant records so their values won't be serialized and thus have to be correctly initialized creating a new instance that you will load the record into.

**Reply by wangjiawen2013** [reply]
```rust
let model: MyModel<B> = MyModel {
    linear: LinearConfig::new(10, 16).init(&device),
    layer: MyLayer {
        l1: LinearConfig::new(16, 8).init(&device),
        l2: LinearConfig::new(8, 2).init(&device),
    },
};

// Apply a function to the weights (here, clamp between -0.5 and 0.5)
model.linear.weight.map(|x| x.clamp(-0.5, 0.5))
```
I tested the above `model.linear.weight.map(|x| x.clamp(-0.5, 0.5)` and find the model was consumed and I cannot see the weight after clamping. It means that we cannot use the modified model later, which is a unexpected behavior.

**Reply by laggui** [reply]
That's my bad, I should have been more clear. `.map(...)` always returns the mapped value. You need to assign the mapped value:

```rust
// Make sure the model is mutable
// let mut model = ...;

// Apply a function to the weights (here, clamp between -0.5 and 0.5)
model.linear.weight = model.linear.weight.map(|x| x.clamp(-0.5, 0.5));
```

**Reply by wangjiawen2013** [reply]
Yes, it works !
Now if I want modify the weight with a tensor/array, how to operate ? I tried `model.linear.weight.val() = my_tensor`, it didn't work. There is no usage scenario for this operation. I am just curious about it and perhaps it will be used somewhere.

**Reply by laggui** [reply]
That's because `param.val()` returns a clone of the tensor held by the parameter, so you're not actually modifying the parameter tensor being held.

You could still use `param.map(...)` to achieve that, for example:

```rust
let value = Tensor::<B, 2>::ones(model.linear.weight.shape(), &device);

// Overwrite weight with value tensor
model.linear.weight = model.linear.weight.map(|_x| value);
```

**Reply by wangjiawen2013** [reply]
Great!
I am implementing a LSTM (Long Short-Term Memory Network), I need to initialize all the weights to 1.0. May be we can use the above method. 
In pytorch, we can use
```python
nn.init.constant_(self.weight_ih.bias[self.hidden_size:2*self.hidden_size], 1.0)
nn.init.constant_(self.weight_hh.bias[self.hidden_size:2*self.hidden_size], 1.0)
```
By the way, we can perform orthogonal initialization using pytorch:
```python
for weight in [self.weight_ih.weight, self.weight_hh.weight]:
            nn.init.orthogonal_(weight) 
```
How to do it with burn ? Does burn have `init.orthogonal()` ?

**Reply by laggui** [reply]
Fyi we have an `Lstm` layer if ever you missed it.

For constant initialization, you can use [`Initializer::Ones`](https://github.com/tracel-ai/burn/blob/main/crates/burn-core/src/nn/norm/batch.rs#L56):
```rust
let gamma = Initializer::Ones.init([num_features], device);
```

`Initializer` supports a couple of different schemes but we don't have orthogonal right now.

**Reply by wangjiawen2013** [reply]
Yes, I missed it. I'll try to use it and compare it with pytorch.

You recommened a github repository on VAE，and burn has its own LSTM and transformer. I think burn is nearly sufficient to our daily work (we're focusing on deep learning in biomedicine) once I can implement Reinforcement Learning (DQN) with burn. If you know where to find DQN project in burn, please send me and it will save me a lot of time. Here is one example of reinforcement learning I find: https://github.com/yunjhongwu/burn-rl-examples

By the way, when will you test the WGAN ?

**Reply by wangjiawen2013** [reply]
Another question, I think we may implement stacked LSTM (multi hidden layers LSTM) by using multiple LSTM in burn, right ?

**Reply by laggui** [reply]
I tested the WGAN this Monday, it works great! Just have some code changes I wanted to make to the structure, but I had to switch task to prepare the v0.16 release.

For RL, I linked this repository from another user on discord not long ago: https://github.com/benbaarber/rl. It looks like it also implements DQN.

**Reply by laggui** [reply]
> Another question, I think we may implement stacked LSTM (multi hidden layers LSTM) by using multiple LSTM in burn, right ?

That's correct! The `Lstm` in Burn is for a single unit. So you can use multiple `Lstm` for your use case. I know that PyTorch defines the LSTM configurable for multiple units/layers. In that case, the hidden states would be the inputs to the other units, and the output should be the hidden states of the last unit.

**Reply by wangjiawen2013** [reply]
Though we can construct a StackedLSTM with LSTM, I cannot find a way to construct biLSTM with StackedLSTM. I think it's more convinence to define a biLSTM configurable for multiple units/layers. Specifically, we can add two parameters to biLSTM to make it more flexible and more convinence, such as `bidirectional` and `num_layers`:

When `bidirectional=true` and `num_layers=2`, it's a biLSTM with StackedLSTM.
When `bidirectional=true` and `num_layers=1`, it's a biLSTM with LSTM.
When `bidirectional=false` and `num_layers=2`, it's a StackedLSTM.
When `bidirectional=false` and `num_layers=1`, it degenerates to a LSTM.

Do you plan to implement the biLSTM like this ? If not, I will try to define one named `LSTMNetwork` and provide it as an burn example. So users can get more kinds of LSTM easily.

**Reply by laggui** [reply]
I think I would prefer keeping the current implementations as one unit and offer a `StackedLstm` (or `StackedRnn` of sort, since we also have `Gru`) to configure the number of layers.

An LSTM example for this would be great 😄

**Reply by wangjiawen2013** [reply]
There is also a lack of some advanced features in current LSTM, such as Layer Normalization for stable training and Dropout regularization between layers to alleviate overfitting and Combined weight matrices for efficiency. I'll try to add these features. Then we can discuss about the example and decide how to make it better.

**Reply by laggui** [reply]
Dropout and LayerNorm are separate modules and applied between Lstm layers usually. So that would be in the stacked implementation 🤔

**Reply by wangjiawen2013** [reply]
> That's my bad, I should have been more clear. `.map(...)` always returns the mapped value. You need to assign the mapped value:
> 
> ```rust
> // Make sure the model is mutable
> // let mut model = ...;
> 
> // Apply a function to the weights (here, clamp between -0.5 and 0.5)
> model.linear.weight = model.linear.weight.map(|x| x.clamp(-0.5, 0.5));
> ```

Things get a bit tricky when I overwrite the bias because the bias is `Some(Param)`, which is different from that of weight. This is how I did:
```rust
let bias = Tensor::<NdArray, 1>::ones([16], &device);
let bias = Param::initialized(model.linear.bias.clone().unwrap().id, bias);
model.linear.bias = model.linear.bias.map(|_x| bias);
```
Does it make sense, or do you have an easy way ?

**Reply by laggui** [reply]
Since the bias is wrapped in `Option<...>`, the first `.map(...)` acts on the inner value (the tensor param). So you can do two maps, for example:

```rust
model.linear.bias = model.linear.bias.map(|x| x.map(|x_i| Tensor::ones_like(&x_i)));
```

**Reply by wangjiawen2013** [reply]
I have a sophisticated implementation of Long Short-Term Memory (LSTM) networks in Burn, featuring state-of-the-art architectural enhancements and optimizations. This implementation includes stacked and bidirectional processing capabilities and advanced regularization techniques (layernorm and dropout) and combined weight matrices for efficiency, making it suitable for both research and production environments. More details can be found at the [PyTorch implementation](https://github.com/shiv08/Advanced-LSTM-Implementation-with-PyTorch). Here is the test result:

![屏幕截图 2025-01-25 163811](https://github.com/user-attachments/assets/084c9197-d31f-4bf5-866f-87c9392b7fdd)


This implementation is complementary to Burn's official LSTM, users can choose either one depends on the project's specific needs. Could I pull a request to burn's example?

**Reply by laggui** [reply]
This refers to the stacked LSTM previously discussed, right? Where multiple instances of the `Lstm` cell implemented in Burn are used to implement a multi-layer LSTM.

If so, I think this could help a lot of users that want to use this type of network for their task. We can start with a PR for the example, and possibly add it to the official nn modules in Burn if it can be useful.

**Comment by wangjiawen2013** [comment]
l read carefully on burn's docs and source code. And here is my suggestions:
1. There are too few tutorials on `visitor & mapper` in the burn's book (https://burn.dev/burn-book/building-blocks/module.html#visitor--mapper).
2. More advanced knowledge on rust language is needed (macros, closures), so it's hard for some users to understand the source code. For example, I cannot understand the definition of  `num_params` and cannot even find the definition of `map`, I only find the signature of `map` (https://github.com/tracel-ai/burn/blob/main/crates/burn-core/src/module/base.rs). All these are necessary to operate on the parameters.
That's why I meet so much difficulties on how to operate on the parametes. We're not as familiar with burn as you developers. So more tutorial or annotation is prefered in burn's book and source code, then we can find what we need easily (such as the way to get all the paramids).

**Reply by laggui** [reply]
1. That's a good point! Would adding the example I posted above to the book help?
2. Ahhh yes I noticed that the book did not provide the up to date trait, fixed it in PR #2609 😅

**Reply by wangjiawen2013** [reply]
Good!
